CDCgov/prime-reportstream
# 10706 Enable Authorization Methods on RS API Docs Swagger UI

This PR:

Enabled Authorization on API Docs swagger UI, the supported authorizations:

OAUTH2
Server2Server
Azure API Key

Test Steps:
1. Checkout the PR to local machine and start report stream server as per README
2. At the completion of the RS server start up, the swagger UI of API docs should be provisioned
3. Go to Azure Storage Explorer and copy the URL of swaggerUI index.html, on local setup, it is:
http://127.0.0.1:10000/devstoreaccount1/apidocs/index.html
4. Paste it into a browser, and you will see:
5. On top of the page: there is external Docs link - currently I put Report Stream API Programmer's Guide link there ( can also use the new UP documents "Swagger UI" mark down as the entry point )
6. Click "Authorize" button to bring up authorization panel where the following auth methods are available:
7. "primeSecurityServerToServer  (http, Bearer)" Method 1 - the process of obtaining the token using primeCLI commands or the python sample program remain as the recommended practice: just keep the legacy process since it worked well and mostly is only used by internal developers
8. "primeSecurityAPIKey (apiKey)" Method 2 - Only available on Azure
9. "primeSecurityOAUTH (OAuth2, authorizationCode)" - associated with OKTA Staging (hhs-prime.oktapreview.com)
10. For OAUTH2 to work, the swagger UI needs to be in a browser started with CORS disabled
e.g on Mac, start chrome as below:
`open -n -a /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --args --user-data-dir="/tmp/chrome_dev_test" --disable-web-security`
Note, this PR mostly deal with swagger UI API Docs on a local setup, on PROD and LLEs, there is a ticket #10621 

On auth server side add the redirect url to the redirect URI list.

For local dev test:

`http://127.0.0.1:10000/devstoreaccount1/apidocs/oauth2-redirect.html`


## Changes
- *Added swagger annotations to enable authorizations supported: server 2 server, azure API key*


## Checklist

### Testing
- [X] Added tests: manual tests: see the setup notes above, after Okta login close the swagger UI auth dialogue and call API  end points, e.g. GET public key for an org, to verify access token is checked, either turn on DEBUG logging and check the logging showing Oauth authentication passed, or put a break point at the authentication point in the end point:
e.g. AuthenticatedClaims.authenticate(request) <=== verify the header has "Authorization bearer: ....." with token value


### Process

## Linked Issues
- #10706

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue
#10862 
#10621

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
